src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,290 @@
+        }
+        auto new_segment = std::make_shared<ReferenceSegment>(table_ptr, col_id, pos_list);
+        out_chunk_ptr->add_segment(new_segment);
+        return;
+      }
+
+      // case 2: segment is reference segment
+      // In this case we need to create a new reference segment that points to the table
+      // that the existing reference segment points to. This is needed to keep the number
+      // of indirections low.
+      const auto ref_segment_ptr = std::dynamic_pointer_cast<ReferenceSegment>(segment_ptr);
+      if (ref_segment_ptr) {
+        auto referenced_table = ref_segment_ptr->referenced_table();
+        auto referenced_column_id = ref_segment_ptr->referenced_column_id();
+        auto pos_list = ref_segment_ptr->pos_list();
+        auto filtered_pos_list = std::make_shared<PosList>();
+        for (const auto& pos : *include_rows_ptr) {
+          filtered_pos_list->emplace_back((*pos_list)[pos]);
+        }
+        auto new_segment =
+            std::make_shared<ReferenceSegment>(referenced_table, referenced_column_id, filtered_pos_list);
+        out_chunk_ptr->add_segment(new_segment);
+        return;
+      }
+
+      // we do not support any segment types beyond value, dict, and reference segments
+      throw std::runtime_error("unrecognized segment class at chunk id " + std::to_string(chunk_id) +
+                               " and column id " + std::to_string(_column_id));
+    });
+  }
+  return out_chunk_ptr;
+}
+
+template <typename T>
+std::shared_ptr<std::vector<ChunkOffset>> TableScan::scan_segment(
+    const std::shared_ptr<const ValueSegment<T>> segment_ptr) const {
+  // determine which values match the filter condition in a
+  // value segment.
+  auto include_rows_ptr = std::make_shared<std::vector<ChunkOffset>>();
+  auto values = segment_ptr->values();
+  auto n_values = values.size();
+  for (auto offset = ChunkOffset{0}; offset < n_values; ++offset) {
+    auto value = values[offset];
+    if (matches_search_value<T>(value)) {
+      include_rows_ptr->emplace_back(offset);
+    }
+  }
+  return include_rows_ptr;
+}
+
+template <typename T>
+std::shared_ptr<std::vector<ChunkOffset>> TableScan::scan_segment(
+    const std::shared_ptr<const DictionarySegment<T>> segment_ptr) const {
+  // determine which values match the filter condition in a
+  // dictionary segment.
+  // The current implementation uses an unoptimized approach where
+  // we unpack each value in the dictionary segment and check the
+  // condition. We could improve performance by making use of the
+  // ordered dictionary directly.
+  auto include_rows_ptr = std::make_shared<std::vector<ChunkOffset>>();
+  auto dictionary = segment_ptr->dictionary();
+  auto attribute_vector_ptr = segment_ptr->attribute_vector();
+  auto n_values = attribute_vector_ptr->size();
+  for (auto offset = ChunkOffset{0}; offset < n_values; ++offset) {
+    auto value = dictionary[attribute_vector_ptr->get(offset)];
```

Comment:
You could get the ValueID of the filter value once and check if attribute_vector at offset matches the ValueID.